core/modules/file/file.field.inc

```diff
@@ -556,14 +573,55 @@ function file_field_widget_value($element, $input = FALSE, $form_state) {
 
   // Ensure that all the required properties are returned even if empty.
   $return += array(
-    'fid' => 0,
+    'fids' => array(),
     'display' => 1,
     'description' => '',
   );
 
   return $return;
 }
 
+ /**
+ * Validation callback for upload element on file widget. Checks if user has
+ * uploaded more files than allowed.
+ *
+ * This validator is used only when cardinality not set to 1 or unlimited.
+ */
+function file_field_widget_multiple_count_validate($element, &$form_state, $form) {
+  $parents = $element['#parents'];
+  $values = backdrop_array_get_nested_value($form_state['values'], $parents);
+
+  array_pop($parents);
+  $current = count(element_children(backdrop_array_get_nested_value($form, $parents))) - 1;
```

Comment:
Perhaps change this variable name to `$already_uploaded`, and also `$uploaded` to `$newly_uploaded`. Then finally `$count` to `$file_count`.

...that way, this would be more self-explanatory:

```php
$file_count = $already_uploaded + $newly_uploaded;
if ($file_count > $field['cardinality']) {
```